### ISSUE TITLE
FeignAutoConfiguration method called sequence cause NullPointerException

### ISSUE THREAD

**Issue by kk17** [spring-cloud/spring-cloud-openfeign#46]
I have the follow config:

`build.gradle`:

```gradle
compile ('org.springframework.cloud:spring-cloud-starter-openfeign:1.4.4.RELEASE') {
    exclude group: 'org.springframework.cloud', module: 'spring-cloud-starter-netflix-ribbon'
    exclude group: 'org.springframework.cloud', module: 'spring-cloud-starter-netflix-archaius'
    exclude group: 'io.github.openfeign', module: 'feign-hystri'
}
compile 'io.github.openfeign:feign-okhttp:9.5.0'
```

`application.yml`:

```yml
ribbon:
  eureka:
   enabled: false
feign:
  okhttp:
    enabled: true
  client:
    config:
      default:
        connectTimeout: 5000
        readTimeout: 5000
```

When I run the feign cleit, I got the follow exception:

```
java.lang.NullPointerException
    at feign.okhttp.OkHttpClient.execute(OkHttpClient.java:149)
    at feign.SynchronousMethodHandler.executeAndDecode(SynchronousMethodHandler.java:97)
    at feign.SynchronousMethodHandler.invoke(SynchronousMethodHandler.java:76)
    at feign.ReflectiveFeign$FeignInvocationHandler.invoke(ReflectiveFeign.java:103)
    at com.sun.proxy.$Proxy119.identityCheck(Unknown Source)
    at ai.advance.pentagon.feign.GuardianApiFeignClientTest.testIdentityCheck(GuardianApiFeignClientTest.groovy:36)
```
I debug the source code, I think this NullPointerException is caused by [FeignAutoConfiguration.java](https://github.com/spring-cloud/spring-cloud-openfeign/blob/master/spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignAutoConfiguration.java), when method `feignClient` is invoked before method `client`, `okHttpClient` is null.

```java
protected static class OkHttpFeignConfiguration {

    private okhttp3.OkHttpClient okHttpClient;

    ...

    @Bean
    public okhttp3.OkHttpClient client(OkHttpClientFactory httpClientFactory,
                                       ConnectionPool connectionPool, FeignHttpClientProperties httpClientProperties) {
        Boolean followRedirects = httpClientProperties.isFollowRedirects();
        Integer connectTimeout = httpClientProperties.getConnectionTimeout();
        Boolean disableSslValidation = httpClientProperties.isDisableSslValidation();
        this.okHttpClient = httpClientFactory.createBuilder(disableSslValidation).
                connectTimeout(connectTimeout, TimeUnit.MILLISECONDS).
                followRedirects(followRedirects).
                connectionPool(connectionPool).build();
        return this.okHttpClient;
    }

    ...

    @Bean
    @ConditionalOnMissingBean(Client.class)
    public Client feignClient() {
        return new OkHttpClient(this.okHttpClient);
    }
}
```

**Comment by ryanjbaxter** [comment]
You should be using the BOM to manage dependency versions.  Try generating a project via start.spring.io and adding Feign.  Then try this again.

**Comment by kk17** [comment]
I did use BOM to manage dependency versions, I use direct versions here to info the framework versions I used and I don't the bug comes form dependency version management.

**Comment by kk17** [comment]
https://github.com/kk17/feign-okhttp

**Comment by ryanjbaxter** [comment]
Thanks.  Weird how wrapping the client in an `@Service` exposes the bug.  I pushed the same fix to Spring Cloud Netflix 1.4.x so it will be included in both the next Finchley and Edgware releases.
